libraries/botbuilder/src/skills/skillHandler.ts

```diff
@@ -202,71 +208,157 @@ export class SkillHandler extends ChannelServiceHandler {
                     await this.bot.run(context);
                     break;
                 case ActivityTypes.Event:
-                    SkillHandler.applyEventToTurnContextActivity(context, activity);
+                    SkillHandler.applyEventToTurnContextActivity(context, newActivity);
                     await this.bot.run(context);
                     break;
                 default:
-                    resourceResponse = await context.sendActivity(activity);
+                    resourceResponse = await context.sendActivity(newActivity);
                     break;
             }
-        };
-
-        // Add the channel service URL to the trusted services list so we can send messages back.
-        // the service URL for skills is trusted because it is applied based on the original request
-        // received by the root bot.
-        AppCredentials.trustServiceUrl(skillConversationReference.conversationReference.serviceUrl);
-
-        await (this.adapter as BotFrameworkAdapter).continueConversation(
-            skillConversationReference.conversationReference,
-            skillConversationReference.oAuthScope,
-            callback
-        );
+        });
 
         if (!resourceResponse) {
             resourceResponse = { id: uuid() };
         }
+
         return resourceResponse;
     }
+
+    /**
+     *
+     * UpdateActivity() API for Skill.
+     * @remarks
+     * Edit an existing activity.
+     *
+     * Some channels allow you to edit an existing activity to reflect the new
+     * state of a bot conversation.
+     *
+     * For example, you can remove buttons after someone has clicked "Approve" button.
+     * @param claimsIdentity ClaimsIdentity for the bot, should have AudienceClaim, AppIdClaim and ServiceUrlClaim.
+     * @param conversationId Conversation ID.
+     * @param activityId activityId to update.
+     * @param activity replacement Activity.
+     */
+    protected async onUpdateActivity(
+        claimsIdentity: ClaimsIdentity,
+        conversationId: string,
+        activityId: string,
+        activity: Activity
+    ): Promise<ResourceResponse> {
+        await this.continueConversation(claimsIdentity, conversationId, async (adapter, ref, context) => {
+            const newActivity = TurnContext.applyConversationReference(activity, ref.conversationReference);
+
+            context.activity.id = activityId;
+            context.activity.callerId = `${CallerIdConstants.BotToBotPrefix}${JwtTokenValidation.getAppIdFromClaims(
+                claimsIdentity.claims
+            )}`;
+
+            return context.updateActivity(newActivity);
+        });
+
+        // Note: the original activity ID is passed back here to provide "behavioral" parity with the C# SDK. Due to
+        // some inconsistent method signatures, the proper response is not propagated back through `context.updateActivity`
+        // so we have to manually pass this value back.
+        return { id: activityId };
```

Comment:
This is painful but true... Truth be told, I don't understand the scenario where I would need the activity.id of the activity I just updated (when I needed the activity.id to update it). Especially because the activity.id doesn't change on update. But we might as well take it. 😅

Reply:
We should resolve this entirely by making the `updateActivity` method return a response as it does in C#. I'll file a separate issue for that parity work.

Reply:
https://github.com/microsoft/botbuilder-js/issues/2937